tests/dataset_loader.py

```diff
@@ -0,0 +1,177 @@
+# SPDX-FileCopyrightText: Copyright (c) 2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import random
+import typing
+
+import cupy as cp
+import pandas as pd
+
+import cudf as cdf  # rename to avoid clash with property method
+
+from morpheus.io.deserializers import read_file_to_df
+from utils import TEST_DIRS
+
+
+class DatasetLoader:
+    """
+    Helper class for loading and caching test datasets as DataFrames, along with some common manipulation methods.
+
+    Parameters
+    ----------
+    df_type : typing.Literal['cudf', 'pandas']
+        Type of DataFrame to return unless otherwise explicitly specified.
+    """
+
+    __df_cache: typing.Dict[typing.Tuple[typing.Literal['cudf', 'pandas'], str],
+                            typing.Union[cdf.DataFrame, pd.DataFrame]] = {}
+
+    # Values in `__instances` are instances of `DatasetLoader`
+    __instances: typing.Dict[typing.Literal['cudf', 'pandas'], typing.Any] = {}
+
+    def __init__(self, df_type: typing.Literal['cudf', 'pandas']) -> None:
+        self._default_df_type = df_type
+        self.__instances[df_type] = self
+
+    def get_alt_df_type(self, df_type: typing.Literal['cudf', 'pandas']) -> typing.Literal['cudf', 'pandas']:
+        """Returns the other possible df type."""
+        return 'cudf' if df_type == 'pandas' else 'pandas'
+
+    def clear(self):
+        self.__df_cache.clear()
+
+    def get_df(self,
+               file_path: str,
+               df_type: typing.Literal['cudf', 'pandas'] = None) -> typing.Union[cdf.DataFrame, pd.DataFrame]:
+        """
+        Fetch a DataFrame specified from `file_path`. If `file_path` is not an absolute path, it is assumed to be
+        relative to the `test/tests_data` dir. If a DataFrame matching both `file_path` and `df_type` has already been
+        fetched, then a cached copy will be returned. In the event that a DataFrame matching `file_path` but not
+        `df_type` exists in the cache, then the cached copy will be cast to the appropriate type, stored in the cache
+        and then returned
+        """
+        if os.path.abspath(file_path) != os.path.normpath(file_path):
+            full_path = os.path.join(TEST_DIRS.tests_data_dir, file_path)
+        else:
+            full_path = file_path
+
+        if df_type is None:
+            df_type = self._default_df_type
+
```

Comment:
I would ensure everything is absolute and normalized here just so you dont get duplicates for things like `dir1/dir2/../file1.csv` vs `dir1//file1.csv`

```python
full_path = os.path.abspath(file_path)
```